### ISSUE TITLE
query plan sequencing affected with `__typename` in fragment

### ISSUE THREAD

**Issue by abernix** [apollographql/federation#2128]
### Description

Rather than returning deferred chunks over time, the Router is returning all chunks at once at the tail end of the response **only when a `__typename` field is requested** in certain places. **This may just be a general field dependency thin, not just `__typename`**.

### Reproduction Steps

1. Checkout `supergraph-demo-fed2` as of https://github.com/apollographql/supergraph-demo-fed2/tree/91883c79497d5cc793598ab4ba785bd3a62351a8 (`main` is probably fine, now) ~the PR at https://github.com/apollographql/supergraph-demo-fed2/pull/212 (Note it's on another git remote)~ (it's on `main`, now)
2. Run `make docker-down docker-build-subgraphs-no-cache compose-subgraphs-localhost docker-up-subgraphs-localhost`, which will:
    - Shut down some other docker you probably have running on conflicting ports
	- Make sure you build subgraphs fresh without cache
	- Compose those fresh subgraphs into a supergraph
	- Start the subgraphs and detach, leaving you with ports 4001-4005.
4. From `main` on this repository, run the following, replacing `PATH_TO_CHECKOUT_OF_THAT_PR` appropriately:

    ```
    cargo run -- --supergraph PATH_TO_CHECKOUT_OF_THAT_PR/router/custom-main/localhost/supergraph.graphql --config PATH_TO_CHECKOUT_OF_THAT_PR/router/custom-main/localhost/router.yaml --log debug
    ```
5. Run query which shows *proper functionality* where chunks arrive with about a 5 second delay in them (defined in the `variation` resolver)
    ```
     curl --compressed -i -X POST http://localhost:4000/ \
      -H 'accept: multipart/mixed; deferSpec=20220824, application/json' \
      -H 'content-type: application/json' \
      --data-raw '{"operationName":"deferVariation","variables":{},"query":"query deferVariation { allProducts { ...MyFragment @defer sku, id __typename } } fragment MyFragment on Product { variation { name __typename }  }" }'
    ```

    * For visualization purposes, the query here is:

       ```graphql
	  query deferVariation {
	    allProducts {
	      ...MyFragment @defer
	      sku
	      id
	      __typename
	    }
	  }
	  fragment MyFragment on Product {
	    variation {
	      name
	      __typename
	    }
	  }

       ```

    * ... and the query plan for that is:
       ```graphql
	  QueryPlan {
	    Defer {
	      Primary {
	        {
	          allProducts {
	            sku
	            id
	            __typename
	          }
	        }:
	        Fetch(service: "products", id: 0) {
	          {
	            allProducts {
	              __typename
	              ... on Product {
	                __typename
	                id
	              }
	              sku
	              id
	            }
	          }
	        }
	      }, [
	        Deferred(depends: [0], path: "allProducts.@") {
	          {
	            ... on Product {
	              variation {
	                name
	                __typename
	              }
	            }
	          }:
	          Flatten(path: "allProducts.@") {
	            Fetch(service: "products") {
	              {
	                ... on Product {
	                  __typename
	                  id
	                }
	              } =>
	              {
	                ... on Product {
	                  variation {
	                    __typename
	                    name
	                  }
	                }
	              }
	            },
	          }
	        },
	      ]
	    },
	  }
       ```

6. Run query that shows response coming **all at once without a pause** after 5 seconds (incorrect behavior)

    ```
    curl --compressed -i -X POST http://localhost:4000/ \
      -H 'accept: multipart/mixed; deferSpec=20220824, application/json' \
      -H 'content-type: application/json' \
      --data-raw '{"operationName":"deferVariation","variables":{},"query":"query deferVariation { allProducts { ...MyFragment @defer sku id __typename } } fragment MyFragment on Product { variation { name __typename } __typename }"}'
    ```

    * For visualization purposes, the query here is:
      ```graphql
	  query deferVariation {
	    allProducts {
	      ...MyFragment @defer
	      sku
	      id
	      __typename
	    }
	  }
	  fragment MyFragment on Product {
	    variation {
	      name
	      __typename
	    }
	    __typename
	  }
      ```

    * And the query plan is:

      ```graphql
	  QueryPlan {
	    Defer {
	      Primary {
	        {
	          allProducts {
	            sku
	            id
	            __typename
	          }
	        }:
	        Sequence {
	          Fetch(service: "products", id: 0) {
	            {
	              allProducts {
	                __typename
	                ... on Product {
	                  __typename
	                  id
	                }
	                sku
	                id
	              }
	            }
	          },
	          Flatten(path: "allProducts.@") {
	            Fetch(service: "products") {
	              {
	                ... on Product {
	                  __typename
	                  id
	                }
	              } =>
	              {
	                ... on Product {
	                  variation {
	                    __typename
	                    name
	                  }
	                }
	              }
	            },
	          },
	        }
	      }, [
	        Deferred(depends: [0], path: "allProducts.@") {
	          {
	            ... on Product {
	              variation {
	                name
	                __typename
	              }
	              __typename
	            }
	          }:
	          Flatten(path: "allProducts.@") {
	            Fetch(service: "inventory") {
	              {
	                ... on Product {
	                  __typename
	                  id
	                }
	              } =>
	              {
	                ... on Product {
	                  __typename
	                }
	              }
	            },
	          }
	        },
	      ]
	    },
	  }
      ```

**Comment by abernix** [comment]
I'll try to track this landing in the Router here: https://github.com/apollographql/router/issues/1698

**Comment by pcmanus** [comment]
So this is a bug in the query planner code. And it's not really `__typename` specific, but rather an oversight in the defer code that made us lose the defer information in some circumstance (technically, in cases where the query planner had various options to evaluate; which `__typename` happens to create at times (since `__typename` is accessible from any subgraph really) but could be created through `@shareable` field too for instance).  The fix is trivial as pushed #2129 with it (some object has a `clone` method and I forgot to clone the defer information with everything else).
